src/v/ssx/async-clear.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include <seastar/core/future.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+#include <absl/container/flat_hash_map.h>
+
+namespace ssx {
+
+/**
+ * For sufficiently large containers where the element destructors
+ * do some work, it is problematic to spend a long time clearing
+ * the container without yielding to the scheduler.
+ *
+ * This function yields every so often while erasing all elements
+ * in a container.
+ *
+ * The type is specific to absl::flat_hash_map to avoid accidentially
+ * using this function on types where repeatedly erasing from the start is
+ * very expensive, like std::vector.
+ */
+template<typename K, typename V>
+ss::future<> async_clear(absl::flat_hash_map<K, V>& container) {
+    // Below threshold_size, just call clear().
+    // Otherwise yield to the scheduler every `threshold_size` elements
+    constexpr size_t threshold_size = 100;
+
+    if (container.size() < threshold_size) {
+        container.clear();
+        co_return;
+    }
+
+    while (container.size()) {
+        size_t i = 0;
+        auto it = container.begin();
+        while (i < threshold_size && it != container.end()) {
+            ++i;
+            ++it;
+        }
+
+        // Issue multi-element erase calls to give the container a chance
+        // to do any batching/optimization that wouldn't be possible
+        // for N single-element erase calls.
+        container.erase(container.begin(), it);
```

Comment:
Which would make sense, but currently the implementation is:
```c++
  iterator erase(const_iterator first, const_iterator last) {
    while (first != last) {
      erase(first++);
    }
    return last.inner_;
  }
```
So it's just traversing multiple times,